File: data/figures/chapter04/figure_4_2/plot_4_2.m
function plot_4_2(outputFile)
%PLOT_4_2 Reproduce thesis Figure 4.2 from the sibling CSV file.

scriptDir = fileparts(mfilename('fullpath'));
if nargin < 1 || strlength(string(outputFile)) == 0
    outputFile = fullfile(scriptDir, 'plot_4_2.png');
end
outputFile = char(outputFile);
[outputDir, ~, outputExtension] = fileparts(outputFile);
assert(~isempty(outputExtension), ...
    'The output file must include a file-format suffix.');
if ~isempty(outputDir) && ~isfolder(outputDir)
    mkdir(outputDir);
end

data = readtable(fullfile(scriptDir, 'figure_4_2.csv'), ...
    'TextType', 'string', 'VariableNamingRule', 'preserve');
groups = ["FCC", "BCC", "SC"];
rows = zeros(numel(groups), 1);
for k = 1:numel(groups)
    match = find(data.topology == groups(k));
    assert(isscalar(match), 'The CSV must contain one row for %s.', groups(k));
    rows(k) = match;
end
selected = data(rows, :);

mass = selected.two_wing_mass_kg;
compliance = selected.trim_compliance_Nm;
cellSizeMm = selected.cell_size_mm;
r1 = selected.t1_over_a;
caseNumber = selected.case_id;
utilization = [selected.skin_stress_utilization, ...
    selected.beam_stress_utilization];

colors = [0.00, 0.45, 0.70; 0.85, 0.33, 0.10; 0.47, 0.67, 0.19];
fig = figure('Color', 'w', 'Units', 'inches', ...
    'Position', [1, 1, 7.5, 7.4], 'Visible', 'off');
layout = tiledlayout(fig, 2, 1, 'TileSpacing', 'compact', ...
    'Padding', 'compact');

ax1 = nexttile(layout, 1);
hold(ax1, 'on');
for k = 1:numel(groups)
    scatter(ax1, mass(k), compliance(k), 78, colors(k, :), 'filled', ...
        'MarkerEdgeColor', 'k', 'LineWidth', 0.7);
end
grid(ax1, 'on');
box(ax1, 'on');
xlabel(ax1, 'Two-wing mass, kg');
ylabel(ax1, 'Trim compliance, N m');
title(ax1, 'A. Frozen observed representatives');
xlim(ax1, [min(mass) - 1.2, max(mass) + 1.6]);
ylim(ax1, [min(compliance) - 0.13, max(compliance) + 0.18]);

offsets = [0.28, 0.04; 0.22, 0.04; -1.55, -0.09];
for k = 1:numel(groups)
    label = sprintf('%s %d\na = %.2f mm, r_1 = %.3f', ...
        groups(k), caseNumber(k), cellSizeMm(k), r1(k));
    text(ax1, mass(k) + offsets(k, 1), compliance(k) + offsets(k, 2), ...
        label, 'Color', [0.12, 0.12, 0.12], 'FontSize', 9.5, ...
        'VerticalAlignment', 'middle', 'FontName', 'Times New Roman', ...
        'Interpreter', 'tex');
end

ax2 = nexttile(layout, 2);
bars = bar(ax2, utilization, 'grouped');
bars(1).FaceColor = [0.30, 0.30, 0.30];
bars(2).FaceColor = [0.65, 0.65, 0.65];
bars(1).EdgeColor = 'none';
bars(2).EdgeColor = 'none';
hold(ax2, 'on');
yline(ax2, 1, '--', '220 MPa screen', 'Color', [0.75, 0.12, 0.12], ...
    'LineWidth', 1.2, 'LabelHorizontalAlignment', 'right', ...
    'LabelVerticalAlignment', 'bottom', 'FontName', 'Times New Roman', ...
    'Interpreter', 'none');
grid(ax2, 'on');
box(ax2, 'on');
xticks(ax2, 1:numel(groups));
xticklabels(ax2, cellstr(groups(:) + " " + string(caseNumber(:))));
ylabel(ax2, 'Stress utilization, \sigma_{max}/220 MPa', ...
    'Interpreter', 'tex');
title(ax2, 'B. Retained stress-channel comparison');
ylim(ax2, [0, 1.12]);
legend(ax2, {'Shell von Mises', 'CBEAM max. normal'}, ...
    'Location', 'southwest', 'Box', 'off', 'Interpreter', 'none');

allAxes = findall(fig, 'Type', 'axes');
set(allAxes, 'FontName', 'Times New Roman', 'FontSize', 10.5, ...
    'LineWidth', 0.8, 'TickDir', 'out', 'TickLabelInterpreter', 'none');
for k = 1:numel(allAxes)
    allAxes(k).XLabel.Interpreter = 'none';
    allAxes(k).YLabel.Interpreter = 'none';
    allAxes(k).Title.Interpreter = 'none';
end
ax2.YLabel.Interpreter = 'tex';

switch lower(outputExtension)
    case {'.pdf', '.svg'}
        exportgraphics(fig, outputFile, 'ContentType', 'vector');
    case {'.png', '.jpg', '.jpeg', '.tif', '.tiff'}
        exportgraphics(fig, outputFile, 'Resolution', 240);
    otherwise
        close(fig);
        error('Unsupported output format: %s', outputExtension);
end
close(fig);
fprintf('%s\n', outputFile);
end

File: data/figures/chapter04/figure_4_2/plot_4_2.py
"""Reproduce thesis Figure 4.2 from its sibling CSV file."""

from __future__ import annotations

import argparse
import csv
import math
from pathlib import Path

import matplotlib

matplotlib.use("Agg")
import matplotlib.pyplot as plt


SCRIPT_DIR = Path(__file__).resolve().parent
CSV_FILE = SCRIPT_DIR / "figure_4_2.csv"
TOPOLOGY_ORDER = ("FCC", "BCC", "SC")
COLORS = {"FCC": "#0072B2", "BCC": "#D95319", "SC": "#77AC30"}


def read_rows() -> list[dict[str, str]]:
    with CSV_FILE.open(newline="", encoding="utf-8") as handle:
        rows = list(csv.DictReader(handle))
    by_topology = {row["topology"]: row for row in rows}
    if set(by_topology) != set(TOPOLOGY_ORDER) or len(rows) != len(TOPOLOGY_ORDER):
        raise ValueError(f"Expected one row for each of {TOPOLOGY_ORDER}")
    ordered = [by_topology[topology] for topology in TOPOLOGY_ORDER]
    numeric_fields = (
        "cell_size_mm",
        "t1_over_a",
        "two_wing_mass_kg",
        "trim_compliance_Nm",
        "skin_stress_utilization",
        "beam_stress_utilization",
    )
    for row in ordered:
        if not all(math.isfinite(float(row[field])) for field in numeric_fields):
            raise ValueError(f"{row['topology']} contains non-finite plotted data")
    return ordered


def configure_style() -> None:
    plt.rcParams.update(
        {
            "font.family": "Times New Roman",
            "font.size": 10.5,
            "axes.linewidth": 0.8,
            "xtick.direction": "out",
            "ytick.direction": "out",
            "pdf.fonttype": 42,
            "ps.fonttype": 42,
        }
    )


def plot(output_file: Path) -> Path:
    rows = read_rows()
    configure_style()
    output_file.parent.mkdir(parents=True, exist_ok=True)
    if not output_file.suffix:
        raise ValueError("--output must include a file-format suffix")

    mass = [float(row["two_wing_mass_kg"]) for row in rows]
    compliance = [float(row["trim_compliance_Nm"]) for row in rows]
    case_ids = [int(row["case_id"]) for row in rows]
    cell_size_mm = [float(row["cell_size_mm"]) for row in rows]
    t1_over_a = [float(row["t1_over_a"]) for row in rows]
    skin_utilization = [float(row["skin_stress_utilization"]) for row in rows]
    beam_utilization = [float(row["beam_stress_utilization"]) for row in rows]

    figure, axes = plt.subplots(2, 1, figsize=(7.5, 7.4))
    ax1, ax2 = axes

    for index, topology in enumerate(TOPOLOGY_ORDER):
        ax1.scatter(
            mass[index],
            compliance[index],
            s=78,
            color=COLORS[topology],
            edgecolor="black",
            linewidth=0.7,
            zorder=3,
        )
    ax1.grid(True)
    ax1.set_xlabel("Two-wing mass, kg")
    ax1.set_ylabel("Trim compliance, N m")
    ax1.set_title("A. Frozen observed representatives")
    ax1.set_xlim(min(mass) - 1.2, max(mass) + 1.6)
    ax1.set_ylim(min(compliance) - 0.13, max(compliance) + 0.18)

    offsets = ((0.28, 0.04), (0.22, 0.04), (-1.55, -0.09))
    for index, topology in enumerate(TOPOLOGY_ORDER):
        label = (
            f"{topology} {case_ids[index]}\n"
            f"a = {cell_size_mm[index]:.2f} mm, r$_1$ = {t1_over_a[index]:.3f}"
        )
        ax1.text(
            mass[index] + offsets[index][0],
            compliance[index] + offsets[index][1],
            label,
            color="#1f1f1f",
            fontsize=9.5,
            verticalalignment="center",
        )

    x_positions = [1.0, 2.0, 3.0]
    width = 0.30
    ax2.bar(
        [x - width / 2 for x in x_positions],
        skin_utilization,
        width,
        color="#4d4d4d",
        edgecolor="none",
        label="Shell von Mises",
    )
    ax2.bar(
        [x + width / 2 for x in x_positions],
        beam_utilization,
        width,
        color="#a6a6a6",
        edgecolor="none",
        label="CBEAM max. normal",
    )
    ax2.axhline(1.0, color="#BF1E1E", linestyle="--", linewidth=1.2)
    ax2.text(
        3.48,
        1.01,
        "220 MPa screen",
        color="#BF1E1E",
        ha="right",
        va="bottom",
        fontsize=9.5,
    )
    ax2.grid(True, axis="y")
    ax2.set_xticks(x_positions)
    ax2.set_xticklabels(
        [f"{topology} {case_id}" for topology, case_id in zip(TOPOLOGY_ORDER, case_ids)]
    )
    ax2.set_ylabel(r"Stress utilization, $\sigma_{max}/220$ MPa")
    ax2.set_title("B. Retained stress-channel comparison")
    ax2.set_ylim(0.0, 1.12)
    ax2.set_xlim(0.5, 3.5)
    ax2.legend(loc="lower left", frameon=False)

    figure.tight_layout()
    figure.savefig(output_file, dpi=240, bbox_inches="tight")
    plt.close(figure)
    return output_file


def main() -> None:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument(
        "--output",
        type=Path,
        default=SCRIPT_DIR / "plot_4_2.png",
        help="Exact output file path; Matplotlib infers the format from its suffix.",
    )
    args = parser.parse_args()
    output_file = plot(args.output.expanduser().resolve())
    print(output_file)


if __name__ == "__main__":
    main()
